### ISSUE TITLE
What is the inference time of the RTDERT model using ONNXRuntime with GPU

### ISSUE THREAD

**Issue by GLaStu** [olibartfast/object-detection-inference#10]
I'm using ONNXRuntime with RTX 3060 GPU on my local workstation, and the whole inference time is about 100ms.Whether this inference time is normal？The image size is [640,640],batch size =1

**Comment by olibartfast** [comment]
Hi,
what was the inference time when running on CPU? what inference time do you get when you run the rtder python evaluation script on raw model?
I will take a look, maybe it could be a matter of GPU initialization and I could add an optional warmup function when doing an inference on image. Also, do you always get 100ms when you run the inference on a video input source?

**Comment by GLaStu** [comment]
> Hi, what was the inference time when running on CPU? what inference time do you get when you run the rtder python evaluation script on raw model? I will take a look, maybe it could be a matter of GPU initialization and I could add an optional warmup function when doing an inference on image. Also, do you always get 100ms when you run the inference on a video input source?

I didn't use the CPU for inference. When I performed inference using the GPU, I had already done a warmup, and ran inference in a loop many times, with the time basically around 100ms.

**Comment by GLaStu** [comment]
> Hi, what was the inference time when running on CPU? what inference time do you get when you run the rtder python evaluation script on raw model? I will take a look, maybe it could be a matter of GPU initialization and I could add an optional warmup function when doing an inference on image. Also, do you always get 100ms when you run the inference on a video input source?

I didn't use the CPU for inference. When I performed inference using the GPU, I had already done a warmup, and ran inference in a loop many times, with the time basically around 100ms.

**Comment by olibartfast** [comment]
All right thanks, I also own a notebook with a 6GB RTX 3060  I could easily check and compare.

**Comment by olibartfast** [comment]
Sorry, that was a stupid bug, I didn't pass the use-gpu command line flag to the backend setup function. Now in debug I'm measuring about 30ms for preprocess + inference + post

**Comment by GLaStu** [comment]
Thanks

**Comment by GLaStu** [comment]
hello again.using ONNXRuntime with RTX 3060 GPU on my local workstation, and the whole inference time is about 75ms.you said:"about 30ms for preprocess + inference + post",is it used onnx or tensorRTas Backend or tensorRT?

**Comment by GLaStu** [comment]
> Sorry, that was a stupid bug, I didn't pass the use-gpu command line flag to the backend setup function. Now in debug I'm measuring about 30ms for preprocess + inference + post

I can't reproduce your inference time metric, do you use onnxruntime, or tensorrt as backend?

**Comment by olibartfast** [comment]
yes, onnxruntime with gpu support, have you pulled the new code after the fix? 
As soon as I can, I will list here my steps to reproduce it.

**Comment by GLaStu** [comment]
> yes, onnxruntime with gpu support, have you pulled the new code after the fix? As soon as I can, I will list here my steps to reproduce it.

i pulled the new code after the fix. Does this mean that it is possible that the version of ONNX will affect the inference time?My version is 1.13.1-gpu

**Comment by olibartfast** [comment]
I don't think would affect it. Have you passed the cli arg '--use-gpu'? Also try to check passing args --warmup and --benchmark (supposing you are doing inference on an image)

**Comment by GLaStu** [comment]
I think i kown what happend. the program show some info like this: Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.

**Comment by olibartfast** [comment]
Are the weights custom or pre-trained?

**Comment by GLaStu** [comment]
> Are the weights custom or pre-trained?
the weights custom and pre-trained, the time is the same

**Comment by GLaStu** [comment]
> Are the weights custom or pre-trained?

the weights custom and pre-trained, the time is the same

**Comment by olibartfast** [comment]
Can you specify the pre-trained version, the one I used in this project? i.e. rtdetr r18vd?

**Comment by GLaStu** [comment]
D:\YoloDetector\DetectorTHz\x64\Release>RT_DERTv2.exe --type=rtdetrul --source=D:\ultralytics\bus.jpg --labels=D:\YoloDetector\RTDERT\labels.txt --weights=D:\ultralytics\rtdetr-l.onnx --use-gpu --warmup --benchmark
(INFO) Source D:\ultralytics\bus.jpg
(INFO) Weights D:\ultralytics\rtdetr-l.onnx
(INFO) Labels file D:\YoloDetector\RTDERT\labels.txt
(INFO) Detector type rtdetrul
(INFO) Available providers:
(INFO) CUDAExecutionProvider
(INFO) Using CUDA GPU
2024-07-26 14:30:12.8002361 [W:onnxruntime:, session_state.cc:1030 onnxruntime::VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-07-26 14:30:12.8050665 [W:onnxruntime:, session_state.cc:1032 onnxruntime::VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
(INFO) Input Node Name/Shape (1):
(INFO)  images : 1x3x640x640
(INFO)  Data Type: Float
(INFO) channels 3
(INFO) width 640
(INFO) height 640
(INFO) Output Node Name/Shape (1):
(INFO)  output0 : 1x300x84
(INFO) Warmup...
(INFO) preprocess time: 2 ms
(INFO) Inference time: 72 ms
(INFO) postprocess time: 0 ms
(INFO) Iteration 0: 76ms
(INFO) Iteration 1: 77ms
(INFO) Iteration 2: 76ms
(INFO) Iteration 3: 78ms
(INFO) Iteration 4: 77ms
(INFO) Iteration 5: 76ms
(INFO) Iteration 6: 76ms
(INFO) Iteration 7: 76ms
(INFO) Iteration 8: 77ms
(INFO) Iteration 9: 77ms
(INFO) Average inference time over 10 iterations:  76.6ms

**Comment by GLaStu** [comment]
> Ok I see you are testing a different rtdetr version, my previous log was about rtdetr_r18vd, not the ultralytics implementation, i will make a run on that model to compare

Could you tell me the GitHub repository of this rtdert?

**Comment by olibartfast** [comment]
it's all in the readme 
https://github.com/olibartfast/object-detection-inference?tab=readme-ov-file#exporting-a-model-for-inference
Also check this issue #8

**Comment by olibartfast** [comment]
this is my log with rtdetr-l onnx from ultralytics, always running on my gpu  RTX 3060 
 9971 ObjectDetectionApp.cpp:66] Source data/bus.jpg
I0729 10:21:34.908263  9971 ObjectDetectionApp.cpp:67] Weights /home/oli/repos/ultralytics/rtdetr-l.onnx
I0729 10:21:34.908267  9971 ObjectDetectionApp.cpp:68] Labels file /home/oli/repos/object-detection-inference/labels/coco.names
I0729 10:21:34.908270  9971 ObjectDetectionApp.cpp:69] Detector type rtdetrul
I0729 10:21:34.915899  9971 ORTInfer.cpp:12] Available providers:
I0729 10:21:34.915923  9971 ORTInfer.cpp:16] TensorrtExecutionProvider
I0729 10:21:34.915928  9971 ORTInfer.cpp:16] CUDAExecutionProvider
I0729 10:21:34.915933  9971 ORTInfer.cpp:20] Using CUDA GPU
2024-07-29 10:21:36.457397801 [W:onnxruntime:, session_state.cc:1169 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-07-29 10:21:36.457422026 [W:onnxruntime:, session_state.cc:1171 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
I0729 10:21:36.565048  9971 ORTInfer.cpp:51] Input Node Name/Shape (1):
I0729 10:21:36.565158  9971 ORTInfer.cpp:57]    images : 1x3x640x640
I0729 10:21:36.565197  9971 ORTInfer.cpp:76]    Data Type: Float
I0729 10:21:36.565202  9971 ORTInfer.cpp:83] channels 3
I0729 10:21:36.565207  9971 ORTInfer.cpp:84] width 640
I0729 10:21:36.565209  9971 ORTInfer.cpp:85] height 640
I0729 10:21:36.565212  9971 ORTInfer.cpp:88] Output Node Name/Shape (1):
I0729 10:21:36.565220  9971 ORTInfer.cpp:93]    output0 : 1x300x84
I0729 10:21:36.571301  9971 ObjectDetectionApp.cpp:145] Warmup...
I0729 10:21:41.172313  9971 ObjectDetectionApp.cpp:154] Inference time: 54 ms
I0729 10:21:41.261081  9971 ObjectDetectionApp.cpp:111] Iteration 0: 57ms
I0729 10:21:41.315797  9971 ObjectDetectionApp.cpp:111] Iteration 1: 54ms
I0729 10:21:41.370123  9971 ObjectDetectionApp.cpp:111] Iteration 2: 54ms
I0729 10:21:41.425056  9971 ObjectDetectionApp.cpp:111] Iteration 3: 54ms
I0729 10:21:41.479612  9971 ObjectDetectionApp.cpp:111] Iteration 4: 54ms
I0729 10:21:41.535638  9971 ObjectDetectionApp.cpp:111] Iteration 5: 55ms
I0729 10:21:41.589921  9971 ObjectDetectionApp.cpp:111] Iteration 6: 54ms
I0729 10:21:41.644542  9971 ObjectDetectionApp.cpp:111] Iteration 7: 54ms
I0729 10:21:41.699244  9971 ObjectDetectionApp.cpp:111] Iteration 8: 54ms
I0729 10:21:41.754693  9971 ObjectDetectionApp.cpp:111] Iteration 9: 55ms
I0729 10:21:41.754956  9971 ObjectDetectionApp.cpp:115] Average inference time over 10 iterations:  54.5ms

**Comment by GLaStu** [comment]
Thank you so much.my time is around 60ms.
